### PR TITLE
Makes rule ids stable and enables state diffing

### PR DESCRIPTION
**Pull request by hediet**

### CODE REVIEW

**Review by lramos15 — APPROVED** [review]

File: src/grammar/grammar.ts [hunk 1]
```diff
@@ -435,18 +445,30 @@ export class AttributedScopeStack {
...
+	 */
 	private constructor(
 		public readonly parent: AttributedScopeStack | null,
 		public readonly scopePath: ScopeStack,
 		public readonly tokenAttributes: EncodedTokenAttributes
 	) {
 	}
 
+	public toString() {
+		return this.getScopeNames().join('.');
```
Comment (alexdima) [inline src/grammar/grammar.ts:464]: I think spaces would be the natural way to join multiple scope names.

File: src/grammar/grammar.ts [hunk 2]
```diff
@@ -797,8 +851,46 @@ export class StateStack implements StackElementDef {
 		}
 		return false;
 	}
+
+	public toStateStackFrame(): StateStackFrame {
+		return {
+			ruleId: ruleIdToNumber(this.ruleId),
+			enterPos: this._enterPos,
+			anchorPos: this._anchorPos,
```
Comment (alexdima) [inline src/grammar/grammar.ts:859]: It isn't necessary to serialize `enterPos` and `anchorPos`, they are used as "storage" during tokenization of a line, but they don't hold any inherent meaning across lines.

File: src/grammar/grammar.ts [hunk 3]
```diff
@@ -746,16 +799,17 @@ export class StateStack implements StackElementDef {
 			outIndex = this.parent._writeString(res, outIndex);
 		}
 
+		// TODO@alexdima why is this "TODO-"?
```
Comment (alexdima) [inline src/grammar/grammar.ts:802]: I don't remember :)

File: src/grammar/grammar.ts [hunk 4]
```diff
@@ -542,16 +564,35 @@ export class AttributedScopeStack {
...
 /**
  * Represents a "pushed" state on the stack (as a linked list element).
  */
-export class StateStack implements StackElementDef {
+export class StateStackImpl implements StateStack {
 	_stackElementBrand: void = undefined;
 
 	// TODO remove me
-	public static NULL = new StateStack(
+	public static NULL = new StateStackImpl(
```
Comment (alexdima) [inline src/grammar/grammar.ts:595]: We could remove `!` from a few lines below here since now these fields are defined as nullable.

File: src/tests/themedTokenizer.ts [hunk 5]
```diff
@@ -37,7 +38,13 @@ export function tokenizeWithTheme(colorMap: string[], fileContents: string, gram
 				color: foregroundColor
 			};
 		}
-		ruleStack = result.ruleStack;
+
+		if (ruleStack) {
+			const diff = diffStateStacksRefEq(ruleStack, result.ruleStack);
+			ruleStack = applyStateStackDiff(ruleStack, diff);
+		} else {
```
Comment (alexdima) [inline src/tests/themedTokenizer.ts:45]: Is this change intentional? Should we have separate tests for diffing?

**Review by alexdima — APPROVED** [review]
Looks good! Left some minor comments.
